### PR TITLE
Wrapping up possible crashes

### PR DESCRIPTION
**Pull request by phoinic**

### CODE REVIEW

File: src/ahp/prover.rs [hunk 1]
```diff
@@ -296,10 +296,16 @@ impl<F: PrimeField> AHPForR1CS<F> {
                 w
             }
         };
-        let (w_poly, remainder) = w_poly.divide_by_vanishing_poly(&domain_x).unwrap();
-        assert!(remainder.is_zero()); // w_poly is divisible by v_X because we set w = 0 on X.
+        let (w_poly, remainder) = match w_poly.divide_by_vanishing_poly(&domain_x) {
+            Some(v) => v,
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:300]: Let's keep the unwrap and assert here: as stated by the comment, this division mustn't fail and must return a 0 remainder. If it happens, it means there is a software bug

File: src/ahp/prover.rs [hunk 2]
```diff
@@ -334,9 +340,15 @@ impl<F: PrimeField> AHPForR1CS<F> {
 
         let msg = ProverMsg::EmptyMessage;
 
-        assert!(w_poly.degree() <= domain_h.size() - domain_x.size() + zk_bound - 1);
-        assert!(z_a_poly.degree() <= domain_h.size() + zk_bound - 1);
-        assert!(z_b_poly.degree() <= domain_h.size() + zk_bound - 1);
+        if w_poly.degree() > domain_h.size() - domain_x.size() + zk_bound - 1 {
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:343]: Let's restore all the asserts here: they mustn't fail, and if they do a software bug is occurring

File: src/ahp/prover.rs [hunk 3]
```diff
@@ -450,10 +470,12 @@ impl<F: PrimeField> AHPForR1CS<F> {
             &state.domain_x,
             domain_h.clone(),
             r_alpha_x_evals,
-        );
+        )?;
         end_timer!(t_poly_time);
 
-        assert!(t_poly.degree() <= domain_h.size() - 1);
+        if t_poly.degree() > domain_h.size() - 1 {
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:476]: Same comment as above

File: src/ahp/prover.rs [hunk 4]
```diff
@@ -474,7 +499,10 @@ impl<F: PrimeField> AHPForR1CS<F> {
         z_poly.coeffs.par_iter_mut()
             .zip(&x_poly.coeffs)
             .for_each(|(z, x)| *z += x);
-        assert!(z_poly.degree() <= domain_h.size() + zk_bound - 1);
+
+        if z_poly.degree() > domain_h.size() + zk_bound - 1 {
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:503]: Same comment as bove

File: src/ahp/prover.rs [hunk 5]
```diff
@@ -538,8 +570,9 @@ impl<F: PrimeField> AHPForR1CS<F> {
             }
         };
 
-
-        assert!(z_1.degree() <= z_1_degree);
+        if z_1.degree() > z_1_degree {
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:573]: Same comment as above

File: src/ahp/prover.rs [hunk 6]
```diff
@@ -553,7 +586,9 @@ impl<F: PrimeField> AHPForR1CS<F> {
             let size_diff = z_1.coeffs.len() as i32 - domain_h.size() as i32;
 
             // Exclude extraordinary high zk_bound
-            assert!(size_diff >= 0 && size_diff <= domain_h.size() as i32);
+            if size_diff < 0 || size_diff > domain_h.size() as i32 {
+                Err(Error::Other("Extraordinary high zk_bound".to_owned()))?
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:590]: "Unsupported zk bound. Min: 0 Max: {}", domain_h.size()

File: src/ahp/prover.rs [hunk 7]
```diff
@@ -564,7 +599,9 @@ impl<F: PrimeField> AHPForR1CS<F> {
         };
         end_timer!(z_1_g_time);
 
-        assert!(z_1_g.degree() <= z_1_degree);
+        if z_1_g.degree() > z_1_degree {
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:602]: Restore assert

File: src/ahp/prover.rs [hunk 8]
```diff
@@ -573,12 +610,17 @@ impl<F: PrimeField> AHPForR1CS<F> {
         let h_1_time = start_timer!(|| "Compute h_1 poly");
 
         let mut h_1 = &outer_poly + &(&z_1 - &z_1_g);
-        h_1 = h_1.divide_by_vanishing_poly(domain_h).unwrap().0;
+        h_1 = match h_1.divide_by_vanishing_poly(domain_h) {
+            Some(v) => v.0,
+            None => Err(Error::Other("Division by vanishing poly failed".to_owned()))?
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:615]: "Division by vanishing poly failed for h_1"

File: src/ahp/prover.rs [hunk 9]
```diff
@@ -573,12 +610,17 @@ impl<F: PrimeField> AHPForR1CS<F> {
...
+        };
         end_timer!(h_1_time);
 
         let msg = ProverMsg::EmptyMessage;
 
-        assert!(h_1.degree() <= 2 * domain_h.size() - 3 + 2 * zk_bound);
+        if h_1.degree() > 2 * domain_h.size() - 3 + 2 * zk_bound {
+            Err(Error::Other("h_1 degree too big".to_owned()))?
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:622]: Restore assert

File: src/ahp/prover.rs [hunk 10]
```diff
@@ -671,11 +713,13 @@ impl<F: PrimeField> AHPForR1CS<F> {
...
         let (z_2, normalized_v) = BoundaryPolynomial::from_non_coboundary_polynomial_evals(
             EvaluationsOnDomain::from_vec_and_domain(f_vals_on_K, domain_k.clone())
-        ).unwrap();
+        )?;
         let z_2 = z_2.polynomial();
         end_timer!(z_2_time);
 
-        assert!(z_2.degree() <= domain_k.size() - 1);
+        if z_2.degree() > domain_k.size() - 1 {
+            Err(Error::Other("z_2 degree too big".to_owned()))?
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:721]: Restore assert

File: src/ahp/prover.rs [hunk 11]
```diff
@@ -686,7 +730,9 @@ impl<F: PrimeField> AHPForR1CS<F> {
         };
         end_timer!(z_2_g_time);
 
-        assert!(z_2_g.degree() <= domain_k.size() - 1);
+        if z_2_g.degree() > domain_k.size() - 1 {
+            Err(Error::Other("z_2_g degree too big".to_owned()))?
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:734]: Restore assert

File: src/ahp/prover.rs [hunk 12]
```diff
@@ -739,13 +785,15 @@ impl<F: PrimeField> AHPForR1CS<F> {
         let h_2_poly_time = start_timer!(|| "Computing h_2 poly");
         let mut f_poly = &z_2_g - &z_2;
         f_poly.coeffs[0] += normalized_v;
-        let h_2 = (&a_poly - &(&b_poly * &f_poly))
-            .divide_by_vanishing_poly(&domain_k)
-            .unwrap()
-            .0;
+        let h_2 = match (&a_poly - &(&b_poly * &f_poly)).divide_by_vanishing_poly(&domain_k) {
+            Some(v) => v.0,
+            None => Err(Error::Other("Division by vanishing poly failed".to_owned()))?
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:790]: "Division by vanishing poly failed for h_2"

File: src/ahp/prover.rs [hunk 13]
```diff
@@ -739,13 +785,15 @@ impl<F: PrimeField> AHPForR1CS<F> {
...
+        };
         end_timer!(h_2_poly_time);
 
-        assert!(h_2.degree() <= 3 * domain_k.size() - 4);
+        if h_2.degree() > 3 * domain_k.size() - 4 {
+            Err(Error::Other("h_2 degree too big".to_owned()))?
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:795]: Restore assert

File: src/ahp/verifier.rs [hunk 14]
```diff
@@ -56,7 +56,9 @@ impl<F: PrimeField> AHPForR1CS<F> {
             .ok_or(SynthesisError::PolynomialDegreeTooLarge)?;
 
         let alpha: F = fs_rng.squeeze_128_bits_challenge();
-        assert!(!domain_h.evaluate_vanishing_polynomial(alpha).is_zero());
+        if domain_h.evaluate_vanishing_polynomial(alpha).is_zero() {
+            Err(Error::Other("domain_h vanishing poly is zero".to_owned()))?
```
Comment (DanieleDiBenedetto) [inline src/ahp/verifier.rs:60]: "Sampled an alpha challenge belonging to H domain"

File: src/ahp/verifier.rs [hunk 15]
```diff
@@ -85,15 +87,17 @@ impl<F: PrimeField> AHPForR1CS<F> {
     pub fn verifier_second_round<R: FiatShamirRng>(
         mut state: VerifierState<F>,
         fs_rng: &mut R,
-    ) -> (VerifierSecondMsg<F>, VerifierState<F>) {
+    ) -> Result<(VerifierSecondMsg<F>, VerifierState<F>), Error> {
 
         let beta: F = fs_rng.squeeze_128_bits_challenge();
-        assert!(!state.domain_h.evaluate_vanishing_polynomial(beta).is_zero());
+        if state.domain_h.evaluate_vanishing_polynomial(beta).is_zero() {
+            Err(Error::Other("domain_h vanishing poly is zero".to_owned()))?
```
Comment (DanieleDiBenedetto) [inline src/ahp/verifier.rs:94]: "Sampled a beta challenge belonging to H domain"

**Review by DanieleDiBenedetto — CHANGES_REQUESTED** [review]

File: src/ahp/prover.rs [hunk 16]
```diff
@@ -300,6 +300,10 @@ impl<F: PrimeField> AHPForR1CS<F> {
         assert!(remainder.is_zero()); // w_poly is divisible by v_X because we set w = 0 on X.
         end_timer!(w_poly_time);
 
+        if !remainder.is_zero() {
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:303]: Leftover ?

File: src/ahp/prover.rs [hunk 17]
```diff
@@ -553,7 +571,12 @@ impl<F: PrimeField> AHPForR1CS<F> {
             let size_diff = z_1.coeffs.len() as i32 - domain_h.size() as i32;
 
             // Exclude extraordinary high zk_bound
-            assert!(size_diff >= 0 && size_diff <= domain_h.size() as i32);
+            if size_diff < 0 || size_diff > domain_h.size() as i32 {
+                Err(Error::Other(format!(
+                    "Unsupported zk bound. Min: 0 Max: {}",
```
Comment (DanieleDiBenedetto) [inline src/ahp/prover.rs:576]: On second thought, let's just leave "Unsupported zk bound"

File: src/lib.rs [hunk 18]
```diff
@@ -172,7 +172,9 @@ impl<F: PrimeField, PC: PolynomialCommitment<F>, D: Digest> Marlin<F, PC, D> {
         zk_rng: Option<&mut dyn RngCore>,
     ) -> Result<Proof<F, PC>, Error<PC::Error>>
     {
-        assert!(zk_rng.is_some() && zk || zk_rng.is_none() && !zk);
+        if zk_rng.is_some() && !zk || zk_rng.is_none() && zk {
+            Err(Error::Other("ZK parameters incorrect".to_owned()))?
```
Comment (DanieleDiBenedetto) [inline src/lib.rs:176]: "If ZK is enabled, a RNG must be passed (and viceversa); conversely, if ZK is disabled, a RNG must NOT be passed (and viceversa)"

**Review by DanieleDiBenedetto — CHANGES_REQUESTED** [review]

**Comment by phoinic** [comment]
Comments from above are addressed

**Review by DanieleDiBenedetto — APPROVED** [review]

**Pull request merged**